lib/ide/src/double_representation/graph.rs

```diff
@@ -68,27 +99,84 @@ pub fn traverse_for_definition
 #[derive(Clone,Debug)]
 pub struct GraphInfo {
     source:DefinitionInfo,
-    /// Describes all known nodes in this graph (does not include special pseudo-nodes like graph
-    /// inputs and outputs).
-    pub nodes:Vec<node::NodeInfo>,
 }
 
 impl GraphInfo {
     /// Describe graph of the given definition.
     pub fn from_definition(source:DefinitionInfo) -> GraphInfo {
-        let nodes = Self::from_function_binding(source.ast.clone());
-        GraphInfo {source,nodes}
+        GraphInfo {source}
     }
 
     /// Lists nodes in the given binding's ast (infix expression).
-    fn from_function_binding(ast:known::Infix) -> Vec<node::NodeInfo> {
+    fn from_function_binding(ast:known::Infix) -> Vec<NodeInfo> {
         let body = ast.rarg.clone();
         if let Ok(body_block) = known::Block::try_new(body.clone()) {
             block_nodes(&body_block)
         } else {
             expression_node(body)
         }
     }
+
+    /// Gets all known nodes in this graph (does not include special pseudo-nodes like graph
+    /// inputs and outputs).
+    pub fn nodes(&self) -> Vec<NodeInfo> {
+        Self::from_function_binding(self.source.ast.clone())
+    }
+
+    /// Adds a new node to this graph.
+    pub fn add_node
+    (&mut self, line_ast:Ast, location_hint:LocationHint) -> FallibleResult<ID> {
+        let block  = self.source.ast.rarg.clone();
+        let block  = known::Block::try_from(block)?;
+
+        let first_line = block.first_line.clone();
+        let mut lines  = block.lines.clone();
+        lines.insert(0, ast::BlockLine{elem:Some(first_line.elem),off:first_line.off});
+
+        let find_position = |id| {
+            lines.iter().find_position(|line| {
+                line.elem.as_ref().map(|line_ast| {
+                    NodeInfo::from_line_ast(line_ast).map(|node| node.id() == id).unwrap_or(false)
+                }).unwrap_or(false)
+            }).map(|(index,_)| index).ok_or(IDNotFound{id})
+        };
+
+        let index = match location_hint {
+            LocationHint::Start      => 0,
+            LocationHint::End        => lines.len(),
+            LocationHint::After(id)  => find_position(id)? + 1,
+            LocationHint::Before(id) => find_position(id)?
+        };
+
+        lines.insert(index, ast::BlockLine { elem: Some(line_ast), off: 0 });
+        let first_line = lines.remove(0);
+        let first_line = ast::BlockLine{elem:first_line.elem.unwrap(),off:first_line.off};
```

Comment:
Similarly, I'd have something like `from_all_lines` or `with_all_lines` that updates lines in block from single vector, while leaving other fields intact. Will be useful when implementing other node APIs.